source/server/admin/prometheus_stats_request.cc

```diff
@@ -0,0 +1,160 @@
+#include "source/server/admin/prometheus_stats_request.h"
+
+#include <string>
+#include <vector>
+
+#include "source/server/admin/prometheus_stats.h"
+
+#include "stats_params.h"
+#include "stats_render.h"
+
+namespace Envoy {
+namespace Server {
+
+namespace {
+bool metricLessThan(const Stats::RefcountPtr<Stats::Metric>& stat1,
+                    const Stats::RefcountPtr<Stats::Metric>& stat2) {
+  ASSERT(&stat1->constSymbolTable() == &stat2->constSymbolTable());
+  return stat1->constSymbolTable().lessThan(stat1->statName(), stat2->statName());
+}
+} // namespace
+
+PrometheusStatsRequest::PrometheusStatsRequest(Stats::Store& stats, const StatsParams& params,
+                                               Stats::CustomStatNamespaces& custom_namespaces,
+                                               UrlHandlerFn url_handler_fn)
+    : StatsRequestBase(stats, params, url_handler_fn), custom_namespaces_(custom_namespaces) {
+
+  // the "type" query param is ignored for prometheus stats, so always start from
+  // counters; also, skip the TextReadouts phase unless that stat type is explicitly
+  // requested via query param
+  if (params_.prometheus_text_readouts_) {
+    phases_ = {Phase{PhaseName::Counters, "Counters"}, Phase{PhaseName::Gauges, "Gauges"},
+               Phase{PhaseName::TextReadouts, "Text Readouts"},
+               Phase{PhaseName::Histograms, "Histograms"}};
+  } else {
+    phases_ = {Phase{PhaseName::Counters, "Counters"}, Phase{PhaseName::Gauges, "Gauges"},
+               Phase{PhaseName::Histograms, "Histograms"}};
+  }
+  phase_ = 0;
+}
+
+template <class StatType> Stats::IterateFn<StatType> PrometheusStatsRequest::checkStat() {
```

Comment:
rename this `saveMatchingStat`? I see the original code used a local lambda `check_stat` for this, but that was mis-named as it's doing more than just checking. Might as well fix it now.